crates/gateway-types/src/class_hash.rs

```diff
@@ -231,6 +280,76 @@ fn compute_class_hash0(mut contract_definition: json::ContractDefinition<'_>) ->
     Ok(ClassHash(outer.finalize()))
 }
 
+/// Computes the class hash for a Sierra class definition.
+///
+/// This matches the (not very precise) [official documentation][starknet-doc] and the [cairo-lang
+/// implementation][cairo-compute] written in Cairo.
+///
+/// Calculation is somewhat simpler than for Cairo classes, since it does _not_ involve serializing
+/// JSON and calculating hashes for the JSON output. Instead, ABI is handled as a string and all other
+/// relevant parts of the class definition are transformed into Felts and hashed using Poseidon.
+///
+/// [starknet-doc]: https://docs.starknet.io/documentation/architecture_and_concepts/Contracts/class-hash/
+/// [cairo-compute]: https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/starknet/core/os/contract_class/contract_class.cairo#L42
+fn compute_sierra_class_hash(
+    contract_definition: json::SierraContractDefinition<'_>,
+) -> Result<ClassHash> {
+    use EntryPointType::*;
+
+    if contract_definition.contract_class_version != "0.1.0" {
+        anyhow::bail!("Unsupported Sierra class version");
+    }
+
+    let mut hash = PoseidonHasher::default();
+
+    const SIERRA_VERSION: Felt = felt_bytes!(b"CONTRACT_CLASS_V0.1.0");
+    hash.write(SIERRA_VERSION.into());
+
+    // It is important process the different entrypoint hashchains in correct order.
```

Comment:
nit, line 174 (223) too

```suggestion
    // It is important to process the different entrypoint hashchains in correct order.
```

Reply:
Thanks, done.